File: Assign/nomode/CTetris.cpp
#include "CTetris.h"
#include<math.h>
Matrix** CTetris::setOfCBlockObjects;
CTetris::~CTetris(){
    for(int i=0;i<nBlockTypes;i++)
        delete [] setOfCBlockObjects[i];
    delete [] setOfCBlockObjects;
}

void CTetris::init(int *setOfBlockArrays[], int MAX_BLK_TYPES, int MAX_BLK_DEGREES){
    Tetris::init(setOfBlockArrays,MAX_BLK_TYPES,MAX_BLK_DEGREES);
    nBlockDegrees=MAX_BLK_DEGREES;
    nBlockTypes=MAX_BLK_TYPES;
    setOfCBlockObjects=new Matrix*[MAX_BLK_TYPES];
    for(int i=0;i<MAX_BLK_TYPES;i++)
    {
        setOfCBlockObjects[i]=new Matrix[MAX_BLK_DEGREES];
        for(int j=0;j<MAX_BLK_DEGREES;j++)
        {
            int len=0;
            int where=0;
            for(int k=0;k!=-1;k=setOfBlockArrays[i*MAX_BLK_DEGREES+j][where++]){
                len++;
                if(k==1) setOfBlockArrays[i*MAX_BLK_DEGREES+j][where-1]=i+1;
            }
            len--;
            len=sqrt(len);
            if(iScreenDw<=len) iScreenDw=len;
            setOfCBlockObjects[i][j]=Matrix(setOfBlockArrays[i*MAX_BLK_DEGREES+j], len, len);
        }
    }
}

TetrisState CTetris::accept(char key){
    state=Running;
    Matrix tempBlk;
    if(key>='0'&&key<='6')
    {
        idxBlockType=key-'0';
        if(justStarted==false)
            deleteFullLines();
        iCScreen=Matrix(oCScreen);
    }
    state=Tetris::accept(key);
    currCBlk=Matrix(setOfCBlockObjects[idxBlockType][idxBlockDegree]);
    tempBlk=iCScreen.clip(top,left,top+currCBlk.get_dy(),left+currCBlk.get_dx());
    tempBlk=tempBlk.add(&currCBlk);
    oCScreen=Matrix(iCScreen);
    oCScreen.paste(&tempBlk,top,left);
    return state;
}
void CTetris::deleteFullLines(){
    for(int y=0;y<iScreenDy;y++) //반복문통해 iScreen의 모든 줄을 돌며 채워진 줄이 있는지 확인
    {
        iScreen=Matrix(oScreen);
        iCScreen=Matrix(oCScreen);
        int top=y;
        int left=iScreenDw;
        bool isDelete=false;
        Matrix tempBlk=iScreen.clip(top,left,top+1,left+iScreenDx); //한줄씩 tempBlk으로 잘라서 확인
        if(tempBlk.sum()==iScreenDx) //한줄에 있는 수들의 합이 iScreenDx와 같으면 채워진거임
            isDelete=true;
        if(isDelete==true) //해당 줄이 채워졌으면
        {
            Matrix CtempBlk; 
            tempBlk=iScreen.clip(0,left,top,left+iScreenDx); 
            //iScreen의 가장 위부터 해당 줄 바로 위까지 잘라서 tempBlk에 저장
            CtempBlk=iCScreen.clip(0,left,top,left+iScreenDx);
            //iCSreen의 가장 위부터 해당 줄 바로 위까지 잘라서 CtempBlk에 저장
            oScreen=Matrix(iScreen);
            oCScreen=Matrix(iCScreen);
            oScreen.paste(&tempBlk,1,left); //oScreen의 두번째 줄이 top이 되도록 tempBlk 붙여넣기
            oCScreen.paste(&CtempBlk,1,left); //oCScreen의 두번째 줄이 top이 되도록 CtempBlk 붙여넣기
        } 
    }
    return;
}

File: mobilePro/src/CTetris.java
public class CTetris extends Tetris{
    private static Matrix[][] setOfCBlockObjects;
    private Matrix iCScreen;
    public Matrix oCScreen;
    private Matrix currCBlk;

    private static Matrix[][] createSetOfCBlocks(int[][][][] setOfArrays) throws Exception {
        int ntypes = setOfArrays.length;
        int ndegrees = setOfArrays[0].length;
        Matrix[][] setOfObjects = new Matrix[nBlockTypes][nBlockDegrees];
        for (int t = 0; t < ntypes; t++)
            for (int d = 0; d < ndegrees; d++)
                setOfObjects[t][d] = new Matrix(setOfArrays[t][d]);
        return setOfObjects;
    }
    public static void init(int[][][][] setOfBlockArrays) throws Exception {
        nBlockTypes = setOfBlockArrays.length;
        nBlockDegrees = setOfBlockArrays[0].length;
        setOfCBlockObjects=createSetOfCBlocks(setOfBlockArrays);
        Tetris.init(setOfBlockArrays);
    }

    public CTetris(int cy, int cx) throws Exception {
        super(cy, cx);
        iCScreen = new Matrix(iScreen);
        oCScreen = new Matrix(oScreen);
    }

    public TetrisState accept(char key) throws Exception {
        Matrix tempCBlk;
        if (state == TetrisState.NewBlock) {
            oCScreen=deleteFullLines(oCScreen, currCBlk, top, iScreenDy, iScreenDx, iScreenDw);
            iCScreen = new Matrix(oCScreen);
        }
        state = super.accept(key);
        currCBlk = setOfCBlockObjects[idxBlockType][idxBlockDegree];
        tempCBlk = iCScreen.clip(top, left, top + currCBlk.get_dy(), left + currCBlk.get_dx());
        tempCBlk = tempCBlk.add(currCBlk);
        oCScreen = new Matrix(iCScreen);
        oCScreen.paste(tempCBlk, top, left); //System.out.println();
        return state;
    }
    private Matrix deleteFullLines(Matrix screen, Matrix blk, int top, int dy, int dx, int dw) throws Exception {
        Matrix line, zero, temp, ctmp;
        if (blk == null) return screen; // called right after the game starts!!
        int cy, y, nDeleted = 0,nScanned = blk.get_dy();
        if (top + blk.get_dy() - 1 >= dy)
            nScanned -= (top + blk.get_dy() - dy);
        zero = new Matrix(1, dx - 2*dw);
        for (y = nScanned - 1; y >= 0 ; y--) {
            cy = top + y + nDeleted;
            line = oScreen.clip(cy, 0, cy + 1, screen.get_dx());
            if (line.sum() == oScreen.get_dx()) {
                temp = oScreen.clip(0, 0, cy, screen.get_dx());
                ctmp=screen.clip(0, 0, cy, screen.get_dx());
                oScreen.paste(temp, 1, 0);
                screen.paste(ctmp, 1, 0);
                oScreen.paste(zero, 0, dw);
                screen.paste(zero, 0, dw);

                nDeleted++;
            }
        }
        return screen;
    }
}